### ISSUE TITLE
Ability to filter access logs

### ISSUE THREAD

**Issue by marco-jantke** [containous/traefik#1752]
### Do you want to request a *feature* or report a *bug*?

**Feature**

### Problem Description

At the moment when AccessLogs are enabled, all requests are logged. It would be good to have a possibility to reduce these logs by some kind of log level, as the amount of logs can grow fast and may not be desired by everyone.

Personally I am interested in logging, when things go wrong. So when it comes to the access logs, I would be interested in requests that produce a 5xx response due to some reason (connection failure to backend, 5xx response from backend).

### Implementation options

It is up for discussion in which way such an option could be configured. 

Some ideas of mine:
- add boolean flag like `errorsOnly=true` 
- allow an expression to specify single status codes or ranges of status codes:
  - `codes: "500"`
  - `codes: "500-599"`
  - `codes: "300, 301, 302, 303, 500-599"`

I actually prefer the first option, as it allows for later extension on what should be logged in a more flexible way.

I also had a look on how haproxy handles this. They in general allow to specify a [log level](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#4.2-log) (e.g. debug, info, error) and then have another option to [raise the log level of requests to errors](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#8.3.3).

**Comment by ldez** [comment]
Every time someone does not respect the issue's template, a little bunnies dies.

![southpark - estrella kill rabbit](https://user-images.githubusercontent.com/5674651/27143914-3cea7896-5130-11e7-81f8-02d8ac2da7a7.gif)
![southpark - estrella kill rabbit2](https://user-images.githubusercontent.com/5674651/27143924-49104c72-5130-11e7-82ae-ccedbd5dfa1c.gif)

Please edit your issue description to respect the [issue template](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE.md) as more as possible.

**Comment by timoreimann** [comment]
Some off-the-top-of-my-head thoughts:

I think we should not try to cover every thinkable corner case (e.g., no "HTTP codes 200 and 400 through 405, unless the response body is non-empty"). IMHO, we should think more in terms of access log "classes" or "types" and output a format that users can filter on according to their needs. In that sense, I agree with @marco-jantke that access log settings expressible by boolean flags can likely be used to correspond to such classes.

Error types definitely sound like a reasonable, first thing to support. HAProxy has this notion of logging _erroneous_ HTTP requests only: They are defined as requests which are either in the 5xx range or had to be retried; an access log record then includes the final HTTP response code and the retry counter. Other than HAProxy, I'd probably make them two distinct settings and allow users to enable one, the other, or both.

**Comment by marco-jantke** [comment]
@ldez better now? 🐼

**Comment by marco-jantke** [comment]
@rjshep have you already had ideas for this when implementing the access logs?

**Comment by marco-jantke** [comment]
@timoreimann I like the idea of having two options, especially as we could implement it step by step. Though I am not sure how the configuration could look like. Something like:

```toml
[accessLog]
  errorsOnly=true
  retriesOnly=true
```

doesn't make too much sense from a semantic point of view. Any suggestion?

At the moment there is btw no information about retries in our logs, this would be an extension that is also required. Once we decide on the design here, we can open new issues to reflect the tasks we have to do.

**Comment by rjshep** [comment]
@marco-jantke This wasn't something I'd really considered as we have a requirement to take all of the access log information.  The changes I added, around how logging is performed and then outputting in JSON gave us what we needed.  However you should be able to build on them for what you're looking at - for example using `[GetLogDataTable](https://github.com/containous/traefik/blob/master/middlewares/accesslog/logger.go#L86)` where you need to add additional info on retries into the [`LogData`](https://github.com/containous/traefik/blob/master/middlewares/accesslog/logdata.go#L119) struct.

For actually controlling what's logged you'll probably want to modify the [CommonLogFormat](https://github.com/containous/traefik/blob/master/middlewares/accesslog/logger_formatters.go#L18) formatter or create a new one if you're intending to add this ability for JSON as we currently use the default logrus formatter.

**Comment by timoreimann** [comment]
@marco-jantke I'd probably gather all access log filters into a dedicated sub-table (in TOML speak), e.g.:

```toml
[accessLog]
  [filters]
    errors=true
    retries=true
```

Apart from the grouping effect, this also has the advantage that it should be easily possible to determine if any filter was enabled by comparing the Go struct the table translates to against the empty struct. If no filter is set, we should access-log everything.

**Comment by emilevauge** [comment]
I'm a bit concerned with this proposal. Nowadays, we tend to push every logs in a centralized logs system and filter on demand:
- we don't want to loose any info. Even if an info is not useful today, it could be useful tomorrow
- storage is now inexpensive
- if you really need this, you can still add a batch filter in your centralized logs system

**Comment by marco-jantke** [comment]
I see it the same as you @emilevauge in an ideal world. Though in scenarios with a lot of traffic and Traefik running in the cloud, this can save some money and resources (IO, Network, CPU of Log System). I opened this proposal of course because it would be quite useful in the environment I want to put Traefik in. But I understand that you will have to maintain the feature from the moment on my future PR is merged. So how to proceed from here on?

**Comment by emilevauge** [comment]
I would suggest to use an external filter (fluentd, Logstash, etc...). Those software have been made to do this well. Good for you ?

**Comment by marco-jantke** [comment]
I will investigate in our setup and have a chat with the persons who know more about our logging setup and will come back next week, whether this is actually a possible for us and if not why.

**Comment by marco-jantke** [comment]
@emilevauge its possible in our infrastructure setup to filter via fluentd. So the issue is solvable for me in my infra without having something to change in Traefik. 

Personally I still think it would be a valuable feature, users will kind of expect of Traefik to have, but as said, thats only my personal view. I am closing the issue now though, in case you decide differently and want to have it included, just reach out to me :)

**Comment by timoreimann** [comment]
Bumping this older issue regarding the possibility to filter access logs, in light of a number of learnings we have made while running Traefik in production.

First of all, the suggestion to store all access logs away and filter at query time turned out to be a complete show stopper for us: We generate over 40 GB of unfiltered logs per hour (or 30 TB per month). Accessing this log volume in a quick (indexed) fashion or paying some SaaS provider to do so for us requires a level of engineering effort / cost that is going to be very hard to sustain.

Consequently, we need to filter logs as they are being generated. Without the help of Traefik, it means we'd have to configure fluentd (our log aggregator) for this purpose. In order to assess the overhead, we patched Traefik to do the same level of filtering in-code so that we could compare the numbers. With only errors (5xx responses, timeouts, retries) being logged, our analysis shows that we require about 100% extra CPU running Traefik without the patch.

Obviously, having to provision twice as many CPUs just because of logging that we mostly throw away later is something we'd rather not have to do, and I can imagine other organizations using Traefik with non-trivial request volumes will feel the same way.

Therefore, I'd like to suggest that we reopen the issue and reconsider the feature request. The patch we applied locally is a handful lines of a production code plus some tests. I know there's always a concern that opening up an area like this in Traefik may attract further feature desires. However, given the numbers I presented above, I believe that allowing the feature represents a fair benefit/cost trade-off while follow-up requests can still be evaluated on a case-by-case basis.

@emilevauge and the rest of @containous/traefik: What do you think?

**Comment by timoreimann** [comment]
@emilevauge awesome :)

if you want me to submit a PR (or paste the diff right here) to better demonstrate what we needed to change in Traefik to make minimal filtering happen, let me know.

**Comment by ldez** [comment]
Closed in favor of the #2875
